### PR TITLE
Laboratory work #1, Ksenia Maiorova - 23FPL2

### PR DESCRIPTION
**Pull request by ksenia-maiorova**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -15,11 +15,22 @@ def tokenize(text: str) -> list[str] | None:
...
+
     Returns:
         list[str] | None: A list of lower-cased tokens without punctuation
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(text, str):
+        return None
+
+    if isinstance(text, str):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:27]: there is no need in this check. Why?

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -34,6 +45,21 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(tokens, list):
+        return None
+
+    if isinstance(tokens, list):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:51]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -34,6 +45,21 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        freq_list = []
+        token_list = []
+        freq_dic = {}
+        for token in tokens:
+            if not isinstance(token, str):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:56]: check it earlier

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -34,6 +45,21 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+                return None
+            if isinstance(token, str):
+                freq_list.append(tokens.count(f"{token}") / len(tokens))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:59]: remove f-string

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -49,6 +75,21 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+    if isinstance(language, str) and isinstance(text, str):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:80]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -49,6 +75,21 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
+        tokens = tokenize(text)
+
+        if not isinstance(tokens, list):
+            return None
+        if isinstance(tokens, list):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:85]: the same about condition

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -49,6 +75,21 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+    if isinstance(language, str) and isinstance(text, str):
+        tokens = tokenize(text)
+
+        if not isinstance(tokens, list):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:83]: check earlier

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -49,6 +75,21 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
+            return None
+        if isinstance(tokens, list):
+            calced_freq = calculate_frequencies(tokens)
+
+            if not isinstance(calced_freq, dict):
+                return None
+            if isinstance(calced_freq, dict):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:90]: .

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -49,6 +75,21 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
+                return dict(zip(['name', 'freq'],
+                                [f'{language}', calced_freq]))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:92]: rm f-string

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -64,6 +105,13 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if (not isinstance(predicted, list) or not isinstance(actual, list)
+            or len(predicted) != len(actual)):
+        return None
+
+    if isinstance(predicted, list) and isinstance(actual, list) and len(predicted) == len(actual):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:112]: the same :)

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -64,6 +105,13 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+        mse = sum((p - a) ** 2 for p, a in zip(predicted, actual)) / len(predicted)
+        return mse
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:114]: return as is

File: lab_1_classify_by_unigrams/settings.json [hunk 12]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 6
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/settings.json:2]: I hope that u can achieve 8 at least! Show me your full power!

File: lab_1_classify_by_unigrams/start.py [hunk 13]
```diff
@@ -14,9 +15,12 @@ def main() -> None:
         de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:17]: revert this deletion

File: lab_1_classify_by_unigrams/start.py [hunk 14]
```diff
@@ -14,9 +15,12 @@ def main() -> None:
...
-    assert result, "Detection result is None"
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:18]: revert this deletion

File: lab_1_classify_by_unigrams/start.py [hunk 15]
```diff
@@ -14,9 +15,12 @@ def main() -> None:
...
+
+    print(tokenize(en_text))
+    print(create_language_profile("en", en_text))
+
 
 
 if __name__ == "__main__":
     main()
+
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:26]: useless whitespace. rm

File: seminars/practice_2_strings.py [hunk 16]
```diff
@@ -53,8 +53,11 @@ def multiply_string(input_string: str, how_many: int) -> str:
     display the given string the number of times given in the `how_many`.
     """
     # student realisation goes here
+    new_string = input_string * how_many
```
Comment (EIzharikov) [inline seminars/practice_2_strings.py:56]: revert this changes

**Review by EIzharikov — COMMENTED** [review]
Nice result! Make CI green and go to 8, you can do it! As for seminar file, you can just copy the alll code of this file in root repo and place in this file. If you will have some troubles wite in chat

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:27]: i felt like it needed it, but i do understand now that it's useless
fixed

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:51]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:59]: done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:92]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:80]: fixed

File: lab_1_classify_by_unigrams/start.py (hunk 13 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/start.py:17]: done

File: lab_1_classify_by_unigrams/start.py (hunk 14 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/start.py:18]: done

File: lab_1_classify_by_unigrams/start.py (hunk 15 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/start.py:26]: done

File: seminars/practice_2_strings.py (hunk 16 above)
Comment (ksenia-maiorova) [reply seminars/practice_2_strings.py:56]: done

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:114]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:112]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:90]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:83]: done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:85]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:56]: done

File: lab_1_classify_by_unigrams/start.py [hunk 17]
```diff
@@ -14,7 +15,21 @@ def main() -> None:
         de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+    print(tokenize(en_text))
+    print(create_language_profile("en", en_text))
+    en_profile = create_language_profile("en", en_text)
+    unk_profile = create_language_profile("unk", unknown_text)
+    de_profile = create_language_profile("de", de_text)
+    if (isinstance(en_profile, dict) and isinstance(de_profile, dict)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:23]: do you trust to your code? why isinstance in start.py?

**Review by EIzharikov — COMMENTED** [review]
Nice progress! Make CI green, fix all comments and be ready for the next review. Good luck!

File: lab_1_classify_by_unigrams/start.py (hunk 17 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/start.py:23]: because mypy thinks i shouldn't trust my code :( i'll try to fix it some other way then

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -84,6 +120,17 @@ def compare_profiles(
...
+            or not all(key in unknown_profile for key in ('name', 'freq'))
+            or not all(key in profile_to_compare for key in ('name', 'freq'))):
+        return None
+    tokens_union = set(profile_to_compare['freq'].keys()).union(unknown_profile['freq'].keys())
+    unknown_freq = [unknown_profile['freq'].get(token, 0) for token in tokens_union]
+    freq_to_compare = [profile_to_compare['freq'].get(token, 0) for token in tokens_union]
+    mse = calculate_mse(freq_to_compare, unknown_freq)
+    if not isinstance(mse, float):
+        return None
+    return round(mse, 3)
```
[the remark this answers is not in the document]
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:133]: because i thought that in the laboratory work documentation it says we need to round our result to 3 decimals after the point. i was wrong
fixed

File: lab_1_classify_by_unigrams/start.py (hunk 17 above)
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/start.py]: fixed

File: lab_1_classify_by_unigrams/main.py [hunk 19]
```diff
@@ -15,11 +16,19 @@ def tokenize(text: str) -> list[str] | None:
     Args:
         text (str): A text
 
+
```
[the remark this answers is not in the document]
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:19]: done

File: lab_1_classify_by_unigrams/main.py [hunk 20]
```diff
@@ -49,6 +69,14 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    if (not isinstance(language, str) or not isinstance(text, str)
+            or not isinstance(tokenize(text), list)):
+        return None
+    calced_freq = calculate_frequencies(tokenize(text))
```
[the remark this answers is not in the document]
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:75]: :((
fixed

File: lab_1_classify_by_unigrams/main.py [hunk 21]
```diff
@@ -135,6 +211,29 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+        return None
+    freq_dic = {}
+    for token in profile['freq']:
+        if len(token) == 1:
+            freq_dic[token] = profile['freq'][token]
+
+    unigrams_list = list(freq_dic)
+    preprocessed_profile = {'name': profile['name'], 'freq': {}}
+    for unigram in unigrams_list:
+        if not unigram.isupper():
```
[the remark this answers is not in the document]
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/main.py:227]: in profiles there are the same letters both in lower and uppercase. at first i check if the unigram is.upper() so that there are no mistakes while calculating frequencies
why not islower()? because there are also unigrams like a hieroglyph and a symbol indicating the power of the number which don't have a lowercase version

File: lab_1_classify_by_unigrams/start.py [hunk 22]
```diff
@@ -14,7 +17,26 @@ def main() -> None:
...
+    paths_to_profiles = ["assets/profiles/de.json", "assets/profiles/en.json",
+                         "assets/profiles/es.json", "assets/profiles/fr.json",
+                         "assets/profiles/it.json", "assets/profiles/ru.json",
+                         "assets/profiles/tr.json"]
+    collected_profiles = collect_profiles(paths_to_profiles)
+    if collected_profiles and unk_profile:
+        detected_profiles = detect_language_advanced(unk_profile, collected_profiles)
+        if detected_profiles:
+            print_report(detected_profiles)
     assert result, "Detection result is None"
```
[the remark this answers is not in the document]
Comment (ksenia-maiorova) [reply lab_1_classify_by_unigrams/start.py:40]: done
